app/store.go

```diff
@@ -0,0 +1,186 @@
+package app
+
+import (
+	"bytes"
+	"fmt"
+	"path"
+	"path/filepath"
+	"strings"
+
+	abci "github.com/tendermint/abci/types"
+	"github.com/tendermint/go-wire"
+	"github.com/tendermint/merkleeyes/iavl"
+	cmn "github.com/tendermint/tmlibs/common"
+	dbm "github.com/tendermint/tmlibs/db"
+	"github.com/tendermint/tmlibs/log"
+
+	"github.com/tendermint/basecoin/state"
+)
+
+// Store contains the merkle tree, and all info to handle abci requests
+type Store struct {
+	state.State
+	height    uint64
+	hash      []byte
+	persisted bool
+
+	logger log.Logger
+}
+
+var stateKey = []byte("merkle:state") // Database key for merkle tree save value db values
+
+// ChainState contains the latest Merkle root hash and the number of times `Commit` has been called
+type ChainState struct {
+	Hash   []byte
+	Height uint64
+}
+
+// NewStore initializes an in-memory IAVLTree, or attempts to load a persistant
+// tree from disk
+func NewStore(dbName string, cacheSize int, logger log.Logger) *Store {
+	// start at 1 so the height returned by query is for the
+	// next block, ie. the one that includes the AppHash for our current state
+	initialHeight := uint64(1)
+
+	// Non-persistent case
+	if dbName == "" {
+		tree := iavl.NewIAVLTree(
+			0,
+			nil,
+		)
+		return &Store{
+			State:  state.NewState(tree, false),
+			height: initialHeight,
+			logger: logger,
+		}
+	}
+
+	// Expand the path fully
+	dbPath, err := filepath.Abs(dbName)
+	if err != nil {
+		panic(fmt.Sprintf("Invalid Database Name: %s", dbName))
```

Comment:
Totally agree...
This code was adapted from another package.  Basically all the parsing code copied.
There are way too many panics all over our codebase... and after fighting against them for a while I get lazy.  Thanks for pushing me to do this properly.